src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,6933 @@
+                0, 0, IOUAmount{tiny * Number{105, -2}}));
+            // The purchase price is still too small to affect the total tokens
+            BEAST_EXPECT(ammAlice.expectBalances(
+                XRP(10000), USD(10000), ammAlice.tokens()));
+        });
+
+        // Reset auth account
+        testAMM([&](AMM& ammAlice, Env& env) {
+            ammAlice.bid(alice, IOUAmount{100}, std::nullopt, {carol});
+            BEAST_EXPECT(ammAlice.expectAuctionSlot({carol}));
+            ammAlice.bid(alice, IOUAmount{100});
+            BEAST_EXPECT(ammAlice.expectAuctionSlot({}));
+            Account bob("bob");
+            Account dan("dan");
+            fund(env, {bob, dan}, XRP(1000));
+            ammAlice.bid(alice, IOUAmount{100}, std::nullopt, {bob, dan});
+            BEAST_EXPECT(ammAlice.expectAuctionSlot({bob, dan}));
+        });
+
+        // Bid all tokens, still own the slot and trade at a discount
+        {
+            Env env(*this);
+            fund(env, gw, {alice, bob}, XRP(2000), {USD(2000)});
+            AMM amm(env, gw, XRP(1000), USD(1010), false, 1000);
+            auto const lpIssue = amm.lptIssue();
+            env.trust(STAmount{lpIssue, 100}, alice);
+            env.trust(STAmount{lpIssue, 50}, bob);
+            env(pay(gw, alice, STAmount{lpIssue, 100}));
+            env(pay(gw, bob, STAmount{lpIssue, 50}));
+            // Alice doesn't have anymore lp tokens
+            amm.bid(alice, 100);
+            // But trades without the fee since she still owns the slot
+            env(pay(alice, bob, USD(10)), path(~USD), sendmax(XRP(10)));
+            BEAST_EXPECT(amm.expectBalances(
+                XRP(1010), USD(1000), IOUAmount{1004887562112089, -9}));
+            // Bob pays the fee
+            env(pay(bob, alice, XRP(10)), path(~XRP), sendmax(USD(11)));
+            BEAST_EXPECT(amm.expectBalances(
+                XRP(1000),
+                STAmount{USD, UINT64_C(101010101010101), -11},
+                IOUAmount{1004887562112089, -9}));
+        }
+    }
+
+    void
+    testInvalidAMMPayment()
+    {
+        testcase("Invalid AMM Payment");
+        using namespace jtx;
+        using namespace std::chrono;
+        using namespace std::literals::chrono_literals;
+
+        // Can't pay into AMM account.
+        // Can't pay out since there is no keys
+        testAMM([&](AMM& ammAlice, Env& env) {
+            env(pay(carol, ammAlice.ammAccount(), XRP(10)),
+                ter(tecNO_PERMISSION));
+            env(pay(carol, ammAlice.ammAccount(), USD(10)),
+                ter(tecNO_PERMISSION));
+        });
+
+        // Can't pay into AMM with escrow.
+        testAMM([&](AMM& ammAlice, Env& env) {
+            auto const seq1 = env.seq(carol);
+            env(escrow(carol, ammAlice.ammAccount(), XRP(1)),
+                condition(cb1),
+                finish_time(env.now() + 1s),
+                fee(1500),
+                ter(tesSUCCESS));
```

Comment:
If an `EscrowCreate` with an AMM destination failed with `tecNO_PERMISSION`, then there would be no way to pay into an AMM with an escrow, because there wouldn't be any escrows.

You could still test that the `EscrowFinish` check works, but you'd have to create the `ltEscrow` by directly calling `OpenLedger::modify()`. Admittedly, that would be inconvenient, but I think that some shenanigans can result if anybody can create an `Escrow` for an AMM. For example, would it be possible to withdraw all and delete the AMM if there's an `Escrow` in its directory?

Reply:
Correction - it's PayChanCreate and not CreateCheck.